# Problem 4

### Largest palindrome product
A palindromic number reads the same both ways. The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 × 99.

Find the largest palindrome made from the product of two 3-digit numbers.

###### Solution

In [1]:
def is_palindrome(n):
    """
    Determines if an integer is a palindrome
    
    Input: int
    
    Return: bool
    """
    digits = []
    while n > 0:
        digits.append(n%10)
        n = n//10
    
    for i in range(len(digits)):
        if digits[i] != digits[-i-1]:
            return False
        
    return True

Testing the function to make sure it works

In [2]:
n = 1274721
print('Is the number {0} a palindrome?\n{1}\n'.format(n,is_palindrome(n)))

n = 58291
print('Is the number {0} a palindrome?\n{1}'.format(n,is_palindrome(n)))

Is the number 1274721 a palindrome?
True

Is the number 58291 a palindrome?
False


###### Brute force version of the function to determine the largest palindrome
Just iterates through all possible numbers checking if the new palindrome is greater than the previous ones found

In [3]:
def largest_palindrome_product(factor_digits):
    """
    Determines the largest palindrome product of two integers with a given number of digits
    
    Input: Number of digits
           type: int
           
    Return:
    - Largest palindrome
    type: int
    - Factor1
    type: int
    - Factor2
    type: int
    """
    max_value = 10 ** factor_digits - 1
    min_value = 10 ** (factor_digits-1)
    
    factor1 = max_value
    min_factor2 = 0
    largest = 0
    
    while factor1 > min_value:
        factor2 = max_value
        while factor2 >= factor1:
            if is_palindrome(factor1 * factor2):
                if factor1 * factor2 > largest:
                    largest = factor1 * factor2
                    f1 = factor1
                    f2 = factor2
            factor2 -= 1
        factor1 -= 1
    
    return largest, f1, f2

In [5]:
import time

t0 = time.time()
largest,f1,f2 = largest_palindrome_product(2)
t1 = time.time()
print("The largest palindrome product of two-digit numbers is: {2} = {0} * {1}".format(f1,f2,largest))
print("Time to run the function was: {}s".format(t1-t0))

t0 = time.time()
largest,f1,f2 = largest_palindrome_product(3)
t1 = time.time()
print("\nThe largest palindrome product of three-digit numbers is: {2} = {0} * {1}".format(f1,f2,largest))
print("Time to run the function was: {}s".format(t1-t0))

t0 = time.time()
largest,f1,f2 = largest_palindrome_product(4)
t1 = time.time()
print("\nThe largest palindrome product of four-digit numbers is: {2} = {0} * {1}".format(f1,f2,largest))
print("Time to run the function was: {}s".format(t1-t0))

The largest palindrome product of two-digit numbers is: 9009 = 91 * 99
Time to run the function was: 0.008002519607543945s

The largest palindrome product of three-digit numbers is: 906609 = 913 * 993
Time to run the function was: 0.7518208026885986s

The largest palindrome product of four-digit numbers is: 99000099 = 9901 * 9999
Time to run the function was: 123.68189740180969s


Although running the function for 3-digit was still fast, there are a few optimizations that can make the function execute much faster by eliminating some iterations that could never result in a greater product than the current highest palindrome

###### Almost the same solution, but a little smarter

In [6]:
def largest_palindrome_product(factor_digits):
    """
    Determines the largest palindrome product of two integers with a given number of digits
    
    Input: Number of digits
           type: int
           
    Return:
    - Largest palindrome
    type: int
    - Factor1
    type: int
    - Factor2
    type: int
    """
    max_value = 10 ** factor_digits - 1
    min_value = 10 ** (factor_digits-1)
    
    factor1 = max_value
    min_factor2 = 0
    largest = 0
    
    while factor1 > max(min_value,largest//max_value):
        factor2 = max_value
        while factor2 >= max(factor1,min_factor2):
            if is_palindrome(factor1 * factor2):
                min_factor2 = factor2
                if factor1 * factor2 > largest:
                    largest = factor1 * factor2
                    f1 = factor1
                    f2 = factor2
            factor2 -= 1
        factor1 -= 1
    
    return largest, f1, f2

In [7]:
import time

t0 = time.time()
largest,f1,f2 = largest_palindrome_product(2)
t1 = time.time()
print("The largest palindrome product of two-digit numbers is: {2} = {0} * {1}".format(f1,f2,largest))
print("Time to run the function was: {}s".format(t1-t0))

t0 = time.time()
largest,f1,f2 = largest_palindrome_product(3)
t1 = time.time()
print("\nThe largest palindrome product of three-digit numbers is: {2} = {0} * {1}".format(f1,f2,largest))
print("Time to run the function was: {}s".format(t1-t0))

t0 = time.time()
largest,f1,f2 = largest_palindrome_product(4)
t1 = time.time()
print("\nThe largest palindrome product of four-digit numbers is: {2} = {0} * {1}".format(f1,f2,largest))
print("Time to run the function was: {}s".format(t1-t0))

t0 = time.time()
largest,f1,f2 = largest_palindrome_product(5)
t1 = time.time()
print("\nThe largest palindrome product of five-digit numbers is: {2} = {0} * {1}".format(f1,f2,largest))
print("Time to run the function was: {}s".format(t1-t0))

t0 = time.time()
largest,f1,f2 = largest_palindrome_product(6)
t1 = time.time()
print("\nThe largest palindrome product of six-digit numbers is: {2} = {0} * {1}".format(f1,f2,largest))
print("Time to run the function was: {}s".format(t1-t0))

t0 = time.time()
largest,f1,f2 = largest_palindrome_product(7)
t1 = time.time()
print("\nThe largest palindrome product of seven-digit numbers is: {2} = {0} * {1}".format(f1,f2,largest))
print("Time to run the function was: {}s".format(t1-t0))

The largest palindrome product of two-digit numbers is: 9009 = 91 * 99
Time to run the function was: 0.0s

The largest palindrome product of three-digit numbers is: 906609 = 913 * 993
Time to run the function was: 0.013527870178222656s

The largest palindrome product of four-digit numbers is: 99000099 = 9901 * 9999
Time to run the function was: 0.02001476287841797s

The largest palindrome product of five-digit numbers is: 9966006699 = 99681 * 99979
Time to run the function was: 0.19617271423339844s

The largest palindrome product of six-digit numbers is: 999000000999 = 999001 * 999999
Time to run the function was: 2.283132553100586s

The largest palindrome product of seven-digit numbers is: 99956644665999 = 9997647 * 9998017
Time to run the function was: 39.20926260948181s


As we can see, the results are the same, but the time it took to run the function for 7-digit numbers was 1/3 of the time the previous function took to run for just 4-digit numbers

THE FINAL ANSWER IS: **906609**